File: Java/TrimABinarySearchTree.java
/**
 * Created by Ethan on 2017/9/18.
 */
public class TrimABinarySearchTree {
    public static TreeNode trimBST(TreeNode root, int L, int R) {
        if (root == null){
            return null;
        }
        if (root.val > R){
            return trimBST(root.left, L, R);
        } else if (root.val < L){
            return trimBST(root.right, L, R);
        } else {
            root.left = trimBST(root.left, L, R);
            root.right = trimBST(root.right, L, R);
            return root;
        }
    }

    public static void main(String[] args) {

    }
}

File: SwiftCode/TrimaBinarySearchTree.swift
//
//  TrimaBinarySearchTree.swift
//  SwiftCode
//
//  Created by Luminoid Li on 1/8/23.
//

import Foundation

//@main
private class Solution {
    // Time complexity: O(n)
    // Space complexity: O(n)
    func trimBST(_ root: TreeNode?, _ low: Int, _ high: Int) -> TreeNode? {
        guard let root = root else { return nil }
        if root.val > high {
            return trimBST(root.left, low, high)
        } else if root.val < low {
            return trimBST(root.right, low, high)
        }
        root.left = trimBST(root.left, low, high)
        root.right = trimBST(root.right, low, high)
        return root
    }
    
    static func main() {
        let solution = Solution()
    }
}
